pkg/pillar/cmd/zedrouter/patchenvelopes.go

```diff
@@ -0,0 +1,331 @@
+// Copyright (c) 2023 Zededa, Inc.
+// SPDX-License-Identifier: Apache-2.0
+
+package zedrouter
+
+import (
+	"encoding/json"
+	"fmt"
+	"sync"
+
+	uuid "github.com/satori/go.uuid"
+
+	"github.com/lf-edge/eve/pkg/pillar/base"
+	"github.com/lf-edge/eve/pkg/pillar/types"
+	"github.com/lf-edge/eve/pkg/pillar/utils/generics"
+)
+
+// PatchEnvelopes is a structure representing
+// Patch Envelopes exposed to App instances via metadata server
+// for more info check docs/PATCH-ENVELOPES.md
+// Must be created by calling NewPatchEnvelopes()
+//
+// Internally, PatchEnvelopes structure stores envelopes which
+// come from EdgeDevConfig parsed by zedagent. This envelopes contains
+// both inline binary artifacts which are ready to be downloaded by app instances
+// and volume references, which are handled by volumemgr.
+// So PatchEnvelopes struct has completedVolumes and contentTreeStatus to store
+// information of all volumes and contentTree handled by volumemgr to link them with
+// patch envelope volume references. ContentTreeStatus is used to retrieve SHA of underlying
+// file.
+// App instances are accessing PatchEnvelopes via metadata server handlers, which is calling
+// PatchEnvelopes.Get() method to get list of available PatchEnvelopeInfo
+// to certain App Instance which are stored in currentState.
+// PatchEnvelopes also hasupdateStateNotificationCh channel
+// to receive notification about the need of updating specified PatchEnvelopes.
+// Those updates are written in envelopesToUpdate and envelopesToDelete boolean maps, that
+// means that request to update same object will be only once, so there will be no queue of
+// go routines piling up but it will take more CPU time to process it.
+// NewPatchEnvelopes() starts goroutine processStateUpdate() which reads from the channel and updates
+// currentState to desired one.
+// This way handlers can do work of determining which patch envelopes actually need change (if any)
+// and send back in go routine rest of the update including slow work.
+// Note that this channels are only accessible from the outside by calling a function which returns
+// write-only channel, meaning that updateStateNotificationCh should not be
+// read from anywhere except processStateUpdate() so that there could not be any deadlock.
+type PatchEnvelopes struct {
+	sync.Mutex
```

Comment:
I don't see where you use this; I assume it needs to be used across all of the maps since they are accessed by different goroutines. It makes sense to document this as well e.g., by a // Protects all of the maps in this struct.

Reply:
Ah - unnamed field hence you use the Lock/Unlock directly. 

But you should still  document what it protects.

And based on my other comments I suspect you need additional locks or use locked maps to make sure that all reads and writes to all maps here are serialized.

Reply:
This Mutex was used to protect maps, I changed maps to locked ones so we don't need this mutex at all, so I removed it